### PR TITLE
fix(bunny): headers property needs to exist

### PR DESCRIPTION
**Pull request by maciekm**
bunny instrumentation errors when rabbitmq messages don't have `headers` property, e.g. published by an app that isn't instrumented yet.

```
E, [2023-09-07T12:01:06.688026 #1] ERROR -- #<Bunny::Session:0x29d88 hidden@hidden:5672, vhost=hidden, addresses=[hidden:5672]>: Uncaught exception from consumer #<Bunny::Consumer:101200 @channel_id=1 @queue=hidden @consumer_tag=hidden>: #<NoMethodError: undefined method `[]' for nil:NilClass 

carrier[key]
^^^^^> @ /gems/ruby/3.1.0/gems/opentelemetry-api-1.2.2/lib/opentelemetry/context/propagation/text_map_getter.rb:16:in `get'
```

you can reproduce the issue by spiting the [example](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/bunny/example/bunny.rb) file into a publisher without instrumentation and a consumer with.

### CODE REVIEW

File: instrumentation/bunny/lib/opentelemetry/instrumentation/bunny/version.rb [hunk 1]
```diff
@@ -7,7 +7,7 @@
 module OpenTelemetry
   module Instrumentation
     module Bunny
-      VERSION = '0.21.0'
```
Comment (arielvalentin) [inline instrumentation/bunny/lib/opentelemetry/instrumentation/bunny/version.rb:10]: Our release processes manages versions. Please revert this change.

**Review by arielvalentin — CHANGES_REQUESTED** [review]

File: instrumentation/bunny/lib/opentelemetry/instrumentation/bunny/patches/channel.rb [hunk 2]
```diff
@@ -18,6 +18,8 @@ def basic_get(queue, opts = { manual_ack: false })
 
               return [delivery_info, properties, payload] unless delivery_info
 
+              properties[:headers] ||= {}
```
[the remark this answers is not in the document]
Comment (maciekm) [reply instrumentation/bunny/lib/opentelemetry/instrumentation/bunny/patches/channel.rb]: @arielvalentin thanks for reviewing, I've now switched to a guard clause where it's being extracted.

still trying to figure out how to test that, any suggestions?

[the remark this answers is not in the document]
Comment (arielvalentin) [reply instrumentation/bunny/lib/opentelemetry/instrumentation/bunny/patches/channel.rb]: I am not familiar with this instrumentation, Bunny, or RabbitMQ so I don't have any advice for you. 

Any suggestions @open-telemetry/ruby-contrib-approvers

File: instrumentation/bunny/test/test_helper.rb [hunk 3]
```diff
@@ -19,4 +19,5 @@
   c.error_handler = ->(exception:, message:) { raise(exception || message) }
   c.logger = Logger.new($stderr, level: ENV.fetch('OTEL_LOG_LEVEL', 'fatal').to_sym)
   c.add_span_processor SPAN_PROCESSOR
+  c.propagators = [OpenTelemetry::Trace::Propagation::TraceContext.text_map_propagator]
```
Comment (maciekm) [inline instrumentation/bunny/test/test_helper.rb:22]: while trying to test I noticed it is only an issue when a single propagator is configured:
```
  1) Error:
OpenTelemetry::Instrumentation::Bunny::PatchHelpers::.extract_context#test_0002_returns the parent context with no links when headers from producer not present:
NoMethodError: undefined method `[]' for nil:NilClass

          carrier[key]
                 ^^^^^
    .../gems/opentelemetry-api-1.2.2/lib/opentelemetry/context/propagation/text_map_getter.rb:16:in `get'
```
when more than one is used, e.g. the default `OTEL_PROPAGATORS=baggage,tracecontext` then it's only a warning:
```
.D, [2023-09-08T12:52:54.081854 #82070] DEBUG -- : Error extracting W3C baggage: undefined method `[]' for nil:NilClass

          carrier[key]
                 ^^^^^
W, [2023-09-08T12:52:54.082129 #82070]  WARN -- : Error in CompositePropagator#extract undefined method `[]' for nil:NilClass

          carrier[key]
                 ^^^^^
```

File: instrumentation/bunny/lib/opentelemetry/instrumentation/bunny/patches/channel.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (maciekm) [reply instrumentation/bunny/lib/opentelemetry/instrumentation/bunny/patches/channel.rb]: @arielvalentin test added

File: instrumentation/bunny/test/test_helper.rb (hunk 3 above)
Comment (arielvalentin) [reply instrumentation/bunny/test/test_helper.rb:22]: @open-telemetry/ruby-maintainers Anything stand out to you about this w/r/t the number of propagators being configured?

Comment (maciekm) [reply instrumentation/bunny/test/test_helper.rb:22]: @arielvalentin could we get this one or https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/664 merged? Both fix the bug.

Comment (arielvalentin) [reply instrumentation/bunny/test/test_helper.rb:22]: I'll give this another look tomorrow. 

I'm not a bunny user so I was trying to get the attention of the original contributors to have a look since they may have additional insights and concerns. 

Thanks again for your contribution and patience.

**Review by arielvalentin — APPROVED** [review]
@maciekm I have not had much success in tracking down or getting feedback from the original contributors.

Would you be willing to take over as the maintainer of this instrumentation?

Also would you be able to look into why this instrumentation has multiple sets of Trace Context Propagation Carries in the properties i.e. `headers` and `tracer_receive_headers`?
